TemporaryProjects.15.0/StartPageExtender.cs

```diff
@@ -0,0 +1,162 @@
+    {
+        private readonly HashSet<IVsWindowFrame> trackedFrames = new HashSet<IVsWindowFrame>();
+        private readonly DTE dte;
+
+        public static async Task InitializeAsync(AsyncPackage package)
+        {
+            await package.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+            var dte = (DTE)await package.GetServiceAsync(typeof(DTE));
+            var vsUIShell = (IVsUIShell)await package.GetServiceAsync(typeof(SVsUIShell));
+
+            ((IVsUIShell7)vsUIShell).AdviseWindowFrameEvents(new StartPageExtender(dte, vsUIShell));
+        }
+
+        private StartPageExtender(DTE dte, IVsUIShell vsUIShell)
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
+
+            this.dte = dte;
+
+            if (ErrorHandler.Succeeded(vsUIShell.FindToolWindow(
+                    (uint)__VSFINDTOOLWIN.FTW_fFrameOnly, VSConstants.StandardToolWindows.StartPage, out var windowFrame)) &&
+                windowFrame != null)
+            {
+                TrackFrame(windowFrame);
+            }
+        }
+
+        public void OnFrameCreated(IVsWindowFrame frame)
+        {
+        }
+
+        public void OnFrameDestroyed(IVsWindowFrame frame)
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
+
+            trackedFrames.Remove(frame);
+        }
+
+        public void OnFrameIsVisibleChanged(IVsWindowFrame frame, bool newIsVisible)
+        {
+        }
+
+        public void OnFrameIsOnScreenChanged(IVsWindowFrame frame, bool newIsOnScreen)
+        {
+        }
+
+        public void OnActiveFrameChanged(IVsWindowFrame oldFrame, IVsWindowFrame newFrame)
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
+
+            if (newFrame != null && oldFrame != newFrame)
+            {
+                TrackFrame(newFrame);
+            }
+        }
+
+        private void TrackFrame(IVsWindowFrame frame)
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
+
+            if (trackedFrames.Add(frame))
+            {
+                if (ErrorHandler.Succeeded(frame.GetGuidProperty((int)__VSFPROPID.VSFPROPID_GuidPersistenceSlot, out var guid)) &&
+                    guid == VSConstants.StandardToolWindows.StartPage &&
+                    ErrorHandler.Succeeded(frame.GetProperty((int)__VSFPROPID.VSFPROPID_DocView, out var windowPane)) &&
+                    // The type of windowPane is Microsoft.VisualStudio.Shell.ToolWindowPane (which inherits from WindowPane),
+                    // but if we tried to cast it, the cast would fail at runtime in VS 2017 where the real type comes from
+                    // Microsoft.VisualStudio.Shell.15.0, whereas we're referencing
+                    // Microsoft.VisualStudio.Shell.14.0.
+                    // Both can't be referenced at the same time so we have to use reflection as a workaround.
+                    (windowPane as dynamic).Content is IVsUIWpfElement wpfElement &&
```

Comment:
We can get rid of the dynamic/reflection hack now! :-) I wasn't sure whether to cast to `ToolWindowPane` or `WindowPane`. I'll leave this up to you.

Reply:
:arrow_right: This is no longer true after 8e79128899b821efb460ccb3b795856d328a88a4

However, I'm not too bothered by that. This solution would not have been very future-proof because the code would have to be duplicated for every new Visual Studio version since each one has a different Shell assembly.